File: 3273-minimum-amount-of-damage-dealt-to-bob/3273-minimum-amount-of-damage-dealt-to-bob.cpp
class Solution {
public:
    long long minDamage(int power, vector<int>& damage, vector<int>& health) {
        #define ld long double
        #define ll long long
        int n=damage.size();
        vector<array<ld, 3>> v(n);
        for(int i=0; i<n; i++){
            int hp = (health[i]+power-1)/power;
            ld dmg = damage[i];
            v[i] = {dmg / (ld)hp, dmg, (ld) hp};
        }
        sort(v.rbegin(), v.rend());
        ll ans=0, T=0;
        for(auto &[_, dmg, hp] : v){
            T += (ll) (hp + 0.00000001);
            ans += (ll) (dmg + 0.00000001) * T;
        }
        return ans;
    }
};

File: 3273-minimum-amount-of-damage-dealt-to-bob/3273-minimum-amount-of-damage-dealt-to-bob.java
class Solution {
    public long minDamage(int power, int[] damage, int[] health) {
        int [][] mArr = new int [damage.length][];
        int val;
        for (int i = 0; i < health.length; ++i) {
            val = health[i] + power - 1;
            val /= power;   
            mArr[i] = new int [] {damage[i], val};
        }
        Arrays.sort(mArr, (a, b) -> a[1]*b[0] - a[0]*b[1]);
        long sum = 0;
        for (int num : damage) 
            sum += num;
        long result = 0;
        for (int [] cArr : mArr) {
            result += sum * cArr[1];
            sum -= cArr[0];
        }
        return result;
    }
}
